src/serverless/structure/serverless_parser.go

```diff
@@ -0,0 +1,317 @@
+	}
+	minResourceLine := math.MaxInt8
+	maxResourceLine := 0
+	for _, funcName := range resourceNames {
+		var existingTags []tags.ITag
+		funcRange := functionsMap[funcName].(map[interface{}]interface{})
+		for key, val := range funcRange {
+			lines := resourceNamesToLines[funcName]
+			switch key {
+			case FunctionTagsAttributeName:
+				funcTags := val.(map[interface{}]interface{})
+				for tagKey, tagValue := range funcTags {
+					existingTags = append(existingTags, &tags.Tag{
+						Key:   tagKey.(string),
+						Value: tagValue.(string),
+					})
+				}
+				tagsLines := p.extractLines(filePath, lines, resourceNames)
+				rawBlock := funcRange
+				minResourceLine = int(math.Min(float64(minResourceLine), float64(lines.Start)))
+				maxResourceLine = int(math.Max(float64(maxResourceLine), float64(lines.End)))
+				slsBlock := &ServerlessBlock{
+					Block: structure.Block{
+						FilePath:          filePath,
+						ExitingTags:       existingTags,
+						RawBlock:          rawBlock,
+						IsTaggable:        true,
+						TagsAttributeName: FunctionTagsAttributeName,
+						Lines:             *lines,
+						TagLines:          tagsLines,
+					},
+					name: funcName,
+				}
+				parsedBlocks = append(parsedBlocks, slsBlock)
+			}
+			p.fileToResourcesLines[filePath] = common.Lines{Start: minResourceLine, End: maxResourceLine}
+		}
+	}
+	return parsedBlocks, nil
+}
+
+func (p *ServerlessParser) extractLines(filePath string, lines *common.Lines, resourceNames []string) common.Lines {
+	tagsLines := p.getTagsLines(filePath, lines, resourceNames)
+	return tagsLines
+}
+
+func (p *ServerlessParser) WriteFile(readFilePath string, blocks []structure.IBlock, writeFilePath string) error {
+	fileFormat := common.GetFileFormat(readFilePath)
+	switch fileFormat {
+	case common.YamlFileType.FileFormat, common.YmlFileType.FileFormat:
+		for _, block := range blocks {
+			serverlessBlock, ok := block.(*ServerlessBlock)
+			if !ok {
+				logger.Warning("failed to convert block to ServerlessBlock")
+				continue
+			}
+			serverlessBlock.UpdateTags()
+		}
+		return structure.WriteYAMLFile(readFilePath, blocks, writeFilePath, p.fileToResourcesLines[readFilePath], FunctionTagsAttributeName)
+
+	default:
+		logger.Warning(fmt.Sprintf("unsupported file type %s", fileFormat))
+		return nil
+	}
+}
+
+func MapResourcesLineYAML(filePath string, resourceNames []string) map[string]*common.Lines {
+	resourceToLines := make(map[string]*common.Lines)
+	computedResources := make(map[string]bool, 0)
+	for _, resourceName := range resourceNames {
+		// initialize a map between resource name and its lines in file
+		resourceToLines[resourceName] = &common.Lines{Start: -1, End: -1}
+	}
+	// #nosec G304
+	file, err := os.Open(filePath)
+	if err != nil {
+		logger.Warning(fmt.Sprintf("failed to read file %s", filePath))
+		return nil
+	}
+	scanner := bufio.NewScanner(file)
+	defer func() {
+		_ = file.Close()
+	}()
```

Comment:
@tronxd It's duplicated code (from the #nosec comment). Should be a common util function `get_file_scanner`

Reply:
Redundant in my opinion. it's a one liner

Reply:
I meant this entire block:

```
// #nosec G304
	file, err := os.Open(filePath)
	if err != nil {
		logger.Warning(fmt.Sprintf("failed to read file %s", filePath))
		return nil
	}
	scanner := bufio.NewScanner(file)
	defer func() {
		_ = file.Close()
	}()
```